processor/blockprocessor/initialize.go

```diff
@@ -0,0 +1,228 @@
+package blockprocessor
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/rpcs"
+
+	"github.com/algorand/indexer/fetcher"
+	"github.com/algorand/indexer/idb"
+	"github.com/algorand/indexer/processor"
+	"github.com/algorand/indexer/processor/blockprocessor/internal"
+)
+
+// InitializeLedger will initialize a ledger to the directory given by the
+// IndexerDbOpts.
+// nextRound - next round to process after initializing.
+// catchpoint - if provided, attempt to use fast catchup.
+func InitializeLedger(ctx context.Context, logger *log.Logger, catchpoint string, nextRound uint64, genesis bookkeeping.Genesis, opts *idb.IndexerDbOptions) error {
+	if nextRound > 0 {
+		if catchpoint != "" {
+			round, _, err := ledgercore.ParseCatchpointLabel(catchpoint)
+			if err != nil {
+				return fmt.Errorf("InitializeLedger() label err: %w", err)
+			}
+			if uint64(round) >= nextRound {
+				return fmt.Errorf("invalid catchpoint: catchpoint round %d should not be ahead of target round %d", uint64(round), nextRound-1)
+			}
+			err = InitializeLedgerFastCatchup(ctx, logger, catchpoint, opts.IndexerDatadir, genesis)
+			if err != nil {
+				return fmt.Errorf("InitializeLedger() fast catchup err: %w", err)
+			}
+		}
+		err := InitializeLedgerSimple(ctx, logger, nextRound-1, &genesis, opts)
+		if err != nil {
+			return fmt.Errorf("InitializeLedger() slow catchup err: %w", err)
+		}
+	}
+	return nil
+}
+
+// InitializeLedgerFastCatchup executes the migration core functionality.
+func InitializeLedgerFastCatchup(ctx context.Context, logger *log.Logger, catchpoint, dataDir string, genesis bookkeeping.Genesis) error {
+	if dataDir == "" {
+		return fmt.Errorf("InitializeLedgerFastCatchup() err: indexer data directory missing")
+	}
+
+	err := internal.CatchupServiceCatchup(ctx, logger, catchpoint, dataDir, genesis)
```

Comment:
maybe `util.MakeLedger` should be called in this function and `CatchupServiceCatchup` just accept a ledger object instead its file path?

Reply:
This is sort of temporary code, we'll definitely be shifting it around a bit so that all the init logic is part of the block processor instead of having it buried in a constructor function.